README
NAME
    Cache - Integration of Intersystems Cache Database into Perl

SYNOPSIS
      use Cache qw(:lowlevel :highlevel);
      _CacheEval '$ZV'; print _CacheConvert();
  
    This module and all modules in the Cache::-Domain require a perl that
    has Cache fully embedded. (such a binary is a dual-binary that is a Perl
    and a Cache binary at the same time. Usually there is a softlink (ln -s
    cache cperl) so you don't need to use cache --perl anymore, it works the
    following way:

    use:

      o cache --perl [perl options ]    and you start perl with embedded cache
      o cperl [ perl options ]          and you get perl with embedded cache
      o cache [ cache options ]         and you get cache with embedded perl

    For backward compatibility with older versions of Cache-Perl

      o cache -perl [ perl options ]
  
    is still supported but this feature is deprecated. Use "--perl" instead
    of "-perl".

    Note: Most of this stuff is the low-level Interface, you normally don't
    need it, except maybe CacheEval and CacheExecute.

      o use Cache::ObjectScript     - for embedded ObjectScript support
      o use Cache::Global           - for high-performance global access (bulk support)
      o use Cache::Routine          - for calling routines and functions
      o use Cache::Bind             - for bidirectional binding of COS Variables to Perl variables
      o use Cache::Util             - for utility functions and helpers

DESCRIPTION
     * This module provides full access to most Cache call-in functions.
     * You should not use the call-in function without exactly knowing what you are doing
     * These function are not exported by default and prepended by a underscore (that means internal).
     * All functions are perlified - you pass a single string if Cache expects a counted string
     * You don't need to check for errors. Most functions raise exceptions on error: use eval { }; to catch them
     * Only "A" functions are supported, no "W". "W" is NOT Unicode anyway, Intersystems simply lies to you.

User Interface for Cache Functions
    CacheEval $expr
         Evaluates a ObjectScript expression and returns its result
         Exception: yes
         Note: This function is slow because it has to preserve terminal settings

    CacheExecute $stmt
         Executes a ObjectScript command and returns nothing.
         Exception: yes
         Note: This function is slow because it has to preserve terminal settings
 
Cache Call-In High-Level Functions
    The high-level functions can be imported by: use Cache ':highlevel';

    _CacheAbort [ CACHE_CTRLC | CACHE_RESJOB ]
         See Cache specification.
         Exception: Yes
         Note: Don't use it.

    $ctx = _CacheContext()
         See Cache specification.
         Exception: No

    $value = _CacheConvert()
         See Cache specification.
         Exception: Yes
         Note: This function calls CacheConvert(CACHE_ASTRING, ...)

    $value = _CacheConvert2()
         This routine uses CacheType() to ask for the type of TOS and
         tries to get the value the fastest way possible.

         Exception: Yes

    _CacheCtrl($bitmap)
         See Cache specification
         Exception: Yes

    $converted = _CacheCvtIn($string, $table)
         See Cache specification
         Exception: Yes

    $converted = _CacheCvtOut($string, $table)
         See Cache specification
         Exception: Yes

    _CacheEnd()
         See Cache specification
         Exception: Yes
         Note: You should NEVER EVER call this! even POSIX::_exit(1); is prefered.

    $error = _CacheError()
         See Cache specification
         Exception: Yes (if a double fault happens)
         Note: No need to call this because every error is reported by croak.

    $errorstring = _CacheErrxlate($errornum)
         See Cache specification
         Exception: No (if the call to CacheErrxlate fails, undef is returned)

    _CacheEval $string
         See Cache specification
         Exception: Yes

    _CacheExecute $string
         See Cache specification
         Exception: Yes

    $prompt = _CachePrompt()
         See Cache specification
         Exception: Yes
         Note: Experts call this functions only by accident. :)

    _CacheSignal $number
         See Cache specification
         Exception: Yes
         Note: Think and you will find out that you don't want it in most cases.

    _CacheStart($flags, $timeout, $princin, $princout)
         See Cache specification
         Exception: Yes
         Note: Don't call it. It's already done. Say simply thanks :)

    $type = _CacheType()
         See Cache specification
         Exception: No (ahm, check the return value for errors)

Cache Low-Level Call-In Functions
     The low-level functions can be imported by:
     use Cache ':lowlevel';

    Use it only IF:

      * you know how to use gdb
      * you want to corrupt the database
      * you never use a condom anyway :)
      * you know what gmngen/checksum/mdate is made for :)

    _CacheCloseOref $oref
         See Cache specification
         Exception: Yes

    _CacheDoFun $rflags, $numargs
         See Cache specification
         Exception: Yes, please.

    _CacheDoRtn $rflags, $numargs
         See Cache specification
         Exception: Oui

    _CacheExtFun $rflags, $numargs
         See Cache specification
         Exception: Da

    _CacheGetProperty()
         See Cache specification
         Exception: Yes, sir.

    _CacheGlobalGet $numsubscipt, $die_or_empty
         See Cache specification
         Exception: yup

    _CacheGlobalSet $numsubscript
         See Cache specification
         Exception: yup, on weekends only.

    _CacheIncrementCountOref $oref
         See Cache specification
         Exception: ja

    _CacheInvokeClassMethod $numarg
         See Cache specification
         Exception: si

    _CachePop $arg
         Not implemented
         Exception: yes

    $val = _CachePopDbl()
         See Cache specification
         Exception: yes

    $val = _CachePopInt()
         See Cache specification
         Exception: yes

    $string = _CachePopList()
         Currently not implemented
         Exception: yes

    $oref = _CachePopOref()
         See Cache specification
         Exception: yes

    $str = _CachePopStr()
         See Cache specification
         Exception: yes

    $ptr = CachePopPtr()
         Not Implemented
         Exception: yes

    _CachePushClassMethod $classname, $methodname, [$flag]/
         See Cache specification
         Exception: /bin/true
         Note: flag defaults to 0

    _CachePushDbl $double
         See Cache specification
         Exception: yes

    $rflags = _CachePushFunc $tag, $routine;
         See Cache specification
         Exception: yes

    $rflags = _CachePushFuncX $tag, $offset, $env, $routine;
         See Cache specification
         Exception: yes

    _CachePushGlobal $global
         See Cache specification
         Exception: yes

    _CachePushGlobalX $global, $env
         See Cache specification
         Exception: yes

    _CachePushInt $i
         See Cache specification
         Exception: yes

    _CachePushList $string
         See Cache specification
         Exception: yes

    _CachePushMethod $oref, $methodname, [$flag]
         See Cache specification
         Exception: yes
         Note: $flag defaults to 0

    _CachePushOref $oref
         See Cache specification
         Exception: yes

    _CachePushProperty $oref, $property
         See Cache specification
         Exception: yes

    _CachePushPtr $value
         See Cache specification
         Exception: yes

    $rflags = _CachePushRtn $tag, $routine
         See Cache specification
         Exception: yes

    $rflags = _CachePushRtnX $tag, $offset, $env, $routine
         See Cache specification
         Exception: yes

    _CachePushStr $string
         See Cache specification
         Exception: yes

    _CacheSetProperty()
         See Cache specification
         Exception: yes

    _CacheUnPop()
         See Cache specification
         Exception: yes

SEE ALSO
    Cache::ObjectScript, Cache::Global, Cache::Routine, Cache::Util,
    Cache::Bind.

AUTHOR
     Stefan Traby <stefan@hello-penguin.com>
     http://hello-penguin.com

COPYRIGHT
     Copyright 2001,2003,2004 by KW-Computer Ges.m.b.H Graz, Austria
     Copyright 2001,2002,2003,2004 by Stefan Traby <stefan@hello-penguin.com>

LICENSE
     This module is licenced under LGPL
     (GNU LESSER GENERAL PUBLIC LICENSE)
     see the LICENSE-file in the toplevel directory of this distribution.



Changes
    - Update the VCS repo link to the GitHub one.

2015-02-24
    - Make sure t/file_lru.t skips all tests if Time::HiRes isn't installed.
        - See https://rt.cpan.org/Ticket/Display.html?id=102316
        - Thanks to Otto J. Makela for the report.
    - Bugfix/maintenance release (2.11)

2014-05-13
    - Fix a small grammar in the docs (double "as") reported by pink_mist.
        - Thanks, pink_mist!
    - Another fix for a small typo by pink_mist.
        - Thanks again.
    - Add Cache category of warnings to make sure warnings do not emit
    an error.
        - See https://rt.cpan.org/Public/Bug/Display.html?id=95608
        - Thanks to Gabor Szabo for the report.
    - Bugfix/maintenance release (2.10)

2014-02-02
    - Use Digest::SHA instead of Digest::SHA1
        - Digest::SHA is core and preferred by Debian.
        - https://rt.cpan.org/Ticket/Display.html?id=92658
    - Bugfix/maintenance release (2.09)

2014-01-26
    - Convert from use "warnings::register;" to "use warnings;".
    - Bugfix/maintenance release (2.08)

2014-01-26
    - Add perl to the prereq in META.yml/META.json
        - Kwalitee / CPANTS
    - Bugfix/maintenance release (2.07)

2013-09-13
    - Increase the delay of the cache invalidation in the tests by default to
    avoid test failures due to high load.
        - https://rt.cpan.org/Public/Bug/Display.html?id=27280
        - One can set the PERL_CACHE_PM_TESTING environment variable to a true
        value to set it back at a quicker-to-run 1 second delay.
        - Thanks to ANDK for reporting it.
    - Fix Bug in Cache::Memory regarding Namespaces
        - https://rt.cpan.org/Ticket/Display.html?id=32339 .
        - ->set after ->set_namespace failed.
        - Thanks to justin@techadvise.com for the report and for a proposed
        fix.
    - Get rid of trailing space.
        - Add t/style-trailing-space.t
    - Bugfix/maintenance release (2.06)

2013-09-05
    - Add t/pod.t and got the embedded PODs to validate.
    - Convert the Build system to Module-Build.
    - Fix https://rt.cpan.org/Ticket/Display.html?id=78817
        - Fix a warning for lc called with undef value
        when "-w" is specified - during some tests.
    - Tests fail or get stuck on Windows - use Devel::AssertOS to make sure
    we don't build there.
        - https://rt.cpan.org/Ticket/Display.html?id=81386
        - if a kind soul will fix the tests and/or the code on MSWin, we will
        remove it.
    - Bugfix/maintenance release (2.05)

2006-02-01
	- Bugfix release (2.04)
	- Fix for failure to call load_callback when verify_callback
	  fails the result (credit to Chris Fletcher).

2005-11-08
	- Fix for set_expiry in Cache::Memory (credit to Sean M. Egan).

2005-10-20
	- Bugfix release (2.03)
	- Fix for cache_umask: individual files were not created with correct
	  permissions (credit to Chris Huegle).

2004-03-23
	- Bugfix release (2.02)
	- Update require to 5.006 since 'use warnings' depends on it
	  (credit to Adam Kennedy).
	- Fixed a comparison issue with DB_File, where it can compare undef's.

2003-12-15
	- Fixed the Cache::freeze() shortcut method which wasn't passing
	  arguments to Cache::Entry::freeze() (credit to Ingo Blechschmidt).

2003-08-18
	- Bugfix release (2.01)
	 o Fixed parsing of all digit expiry times
	 o Fixed use of scalar validity in Cache::Memory
	 o Allowed validity to be set on non-existant entry
	   (sets entry data to zero length)
	 o Fixed package name for Cache::Memory::HeapElem
	 o Documentation fixes

2003-07-07
	- Initial release (2.00)


lib/Cache.pm
=head1 NAME

Cache - the Cache interface

=head1 DESCRIPTION

The Cache modules are designed to assist a developer in persisting data for a
specified period of time.  Often these modules are used in web applications to
store data locally to save repeated and redundant expensive calls to remote
machines or databases.

The Cache interface is implemented by derived classes that store cached data
in different manners (such as files on a filesystem, or in memory).

=head1 USAGE

To use the Cache system, a cache implementation must be chosen to suit your
needs.  The most common is Cache::File, which is suitable for sharing data
between multiple invocations and even between concurrent processes.

Using a cache is simple.  Here is some very simple sample code for
instantiating and using a file system based cache.

  use Cache::File;

  my $cache = Cache::File->new( cache_root => '/tmp/cacheroot' );
  my $customer = $cache->get( $name );

  unless ($customer) {
      $customer = get_customer_from_db( $name );
      $cache->set( $name, $customer, '10 minutes' );
  }

  return $customer;

Of course, far more powerful methods are available for accessing cached data.
Also see the TIE INTERFACE below.

=head1 METHODS

=over

=cut
package Cache;

require 5.006;
use strict;
use warnings;
use Carp;
use Date::Parse;

# For registering the 'Cache' category. See:
# https://rt.cpan.org/Public/Bug/Display.html?id=95608
use warnings::register;

use base qw(Tie::Hash);
use fields qw(
        default_expires removal_strategy size_limit
        load_callback validate_callback);

our $VERSION = '2.10';

our $EXPIRES_NOW = 'now';
our $EXPIRES_NEVER = 'never';

# map of expiration formats to their respective time in seconds
my %_Expiration_Units = ( map(($_,             1), qw(s second seconds sec)),
                          map(($_,            60), qw(m minute minutes min)),
                          map(($_,         60*60), qw(h hour hours)),
                          map(($_,      60*60*24), qw(d day days)),
                          map(($_,    60*60*24*7), qw(w week weeks)),
                          map(($_,   60*60*24*30), qw(M month months)),
                          map(($_,  60*60*24*365), qw(y year years)) );


sub new {
    my Cache $self = shift;
    my $args = $#_? { @_ } : shift;

    ref $self or croak 'Must use a subclass of Cache';

    $self->set_default_expires($args->{default_expires});

        # set removal strategy
    my $strategy = $args->{removal_strategy} || 'Cache::RemovalStrategy::LRU';
    unless (ref($strategy)) {
        eval "require $strategy" or die @_;
        $strategy = $strategy->new();
    }
    $self->{removal_strategy} = $strategy;

    # set size limit
    $self->{size_limit} = $args->{size_limit};

    # set load callback
    $self->set_load_callback($args->{load_callback});

    # set load callback
    $self->set_validate_callback($args->{validate_callback});

    return $self;
}

=item my $cache_entry = $c->entry( $key )

Return a 'Cache::Entry' object for the given key.  This object can then be
used to manipulate the cache entry in various ways.  The key can be any scalar
string that will uniquely identify an entry in the cache.

=cut

sub entry;

=item $c->purge()

Remove all expired data from the cache.

=cut

sub purge;

=item $c->clear()

Remove all entries from the cache - regardless of their expiry time.

=cut

sub clear;

=item my $num = $c->count()

Returns the number of entries in the cache.

=cut

sub count;

=item my $size = $c->size()

Returns the size (in bytes) of the cache.

=cut

# if an argument is provided, then the target is the 'shortcut' method set($key)
sub size {
    my Cache $self = shift;
    return @_? $self->entry_size(@_) : $self->cache_size();
}

# implement this method instead
sub cache_size;


=back

=head1 PROPERTIES

When a cache is constructed these properties can be supplied as options to the
new() method.

=over

=item default_expires

The current default expiry time for new entries into the cache.  This property
can also be reset at any time.

 my $time = $c->default_expires();
 $c->set_default_expires( $expiry );

=cut

sub default_expires {
    my Cache $self = shift;
    return Canonicalize_Expiration_Time($self->{default_expires});
}

sub set_default_expires {
    my Cache $self = shift;
    my ($time) = @_;
    # This could be made more efficient by converting to unix time here,
    # except that special handling would be required for relative times.
    # For now default_expires() does all the conversion.
    $self->{default_expires} = $time;
}

=item removal_strategy

The removal strategy object for the cache.  This is used to remove
object from the cache in order to maintain the cache size limit.

When setting the removal strategy in new(), the name of a strategy package or
a blessed strategy object reference should be provided  (in the former case an
object is constructed by calling the new() method of the named package).

The strategies 'Cache::RemovalStrategy::LRU' and
'Cache::RemovalStrategy::FIFO' are available by default.

 my $strategy = $c->removal_strategy();

=cut

sub removal_strategy {
    my Cache $self = shift;
    return $self->{removal_strategy};
}

=item size_limit

The size limit for the cache.

 my $limit = $c->size_limit();

=cut

sub size_limit {
    my Cache $self = shift;
    return $self->{size_limit};
}

=item load_callback

The load callback for the cache.  This may be set to a function that will get
called anytime a 'get' is issued for data that does not exist in the cache.

 my $limit = $c->load_callback();
 $c->set_load_callback($callback_func);

=cut

sub load_callback {
    my Cache $self = shift;
    return $self->{load_callback};
}

sub set_load_callback {
    my Cache $self = shift;
    my ($load_callback) = @_;
    $self->{load_callback} = $load_callback;
}

=item validate_callback

The validate callback for the cache.  This may be set to a function that will
get called anytime a 'get' is issued for data that does not exist in the
cache.

 my $limit = $c->validate_callback();
 $c->set_validate_callback($callback_func);

=cut

sub validate_callback {
    my Cache $self = shift;
    return $self->{validate_callback};
}

sub set_validate_callback {
    my Cache $self = shift;
    my ($validate_callback) = @_;
    $self->{validate_callback} = $validate_callback;
}


=back

=head1 SHORTCUT METHODS

These methods all have counterparts in the Cache::Entry package, but are
provided here as shortcuts.  They all default to just wrappers that do
'$c->entry($key)->method_name()'.  For documentation, please refer to
Cache::Entry.

=over

=item my $bool = $c->exists( $key )

=cut

sub exists {
    my Cache $self = shift;
    my $key = shift;
    return $self->entry($key)->exists();
}

=item $c->set( $key, $data, [ $expiry ] )

=cut

sub set {
    my Cache $self = shift;
    my $key = shift;
    return $self->entry($key)->set(@_);
}

=item my $data = $c->get( $key )

=cut

sub get {
    my Cache $self = shift;
    my $key = shift;
    return $self->entry($key)->get();
}

=item my $data = $c->size( $key )

=cut

# method is called 'entry_size' as the size() method is also a normal Cache
# method for returning the size of the entire cache.  It calls this instead if
# given an argument.
sub entry_size {
    my Cache $self = shift;
    my $key = shift;
    return $self->entry($key)->size();
}

=item $c->remove( $key )

=cut

sub remove {
    my Cache $self = shift;
    my $key = shift;
    return $self->entry($key)->remove();
}

=item $c->expiry( $key )

=cut

sub expiry {
    my Cache $self = shift;
    my $key = shift;
    return $self->entry($key)->expiry();
}
sub get_expiry { shift->expiry(@_); }

=item $c->set_expiry( $key, $time )

=cut

sub set_expiry {
    my Cache $self = shift;
    my $key = shift;
    return $self->entry($key)->set_expiry(@_);
}

=item $c->handle( $key, [$mode, [$expiry] ] )

=cut

sub handle {
    my Cache $self = shift;
    my $key = shift;
    return $self->entry($key)->handle();
}

=item $c->validity( $key )

=cut

sub validity {
    my Cache $self = shift;
    my $key = shift;
    return $self->entry($key)->validity();
}
sub get_validity { shift->validity(@_); }

=item $c->set_validity( $key, $data )

=cut

sub set_validity {
    my Cache $self = shift;
    my $key = shift;
    return $self->entry($key)->set_validity(@_);
}

=item $c->freeze( $key, $data, [ $expiry ] )

=cut

sub freeze {
    my Cache $self = shift;
    my $key = shift;
    return $self->entry($key)->freeze(@_);
}

=item $c->thaw( $key )

=cut

sub thaw {
    my Cache $self = shift;
    my $key = shift;
    return $self->entry($key)->thaw();
}


=back

=head1 TIE INTERFACE

  tie %hash, 'Cache::File', { cache_root => $tempdir };

  $hash{'key'} = 'some data';
  $data = $hash{'key'};

The Cache classes can be used via the tie interface, as shown in the synopsis.
This allows the cache to be accessed via a hash.  All the standard methods
for accessing the hash are supported , with the exception of the 'keys' or
'each' call.

The tie interface is especially useful with the load_callback to automatically
populate the hash.

=head1 REMOVAL STRATEGY METHODS

These methods are only for use internally (by concrete Cache implementations).

These methods define the interface by which the removal strategy object can
manipulate the cache (the Cache is the 'context' of the strategy).  By
default, methods need to be provided to remove the oldest or stalest objects
in the cache - thus allowing support for the default FIFO and LRU removal
strategies.  All derived Cache implementations should support these methods
and may also introduce additional methods (and additional removal strategies
to match).

=over

=item my $size = $c->remove_oldest()

Removes the oldest entry in the cache and returns its size.

=cut

sub remove_oldest;

=item my $size = $c->remove_stalest()

Removes the 'stalest' (least used) object in the cache and returns its
size.

=cut

sub stalest;

=item $c->check_size( $size )

This method isn't actually part of the strategy interface, nor does it need
to be defined by Cache implementations.  Instead it should be called by
implementations whenever the size of the cache increases.  It will take care
of checking the size limit and invoking the removal strategy if required.  The
size argument should be the new size of the cache.

=cut

sub check_size {
    my Cache $self = shift;
    my ($size) = @_;

    defined $self->{size_limit} or return;

    if ($size > $self->{size_limit}) {
        $self->{removal_strategy}->remove_size(
                $self, $size - $self->{size_limit});
    }
}


=back

=head1 UTILITY METHODS

These methods are only for use internally (by concrete Cache implementations).

=over

=item my $time = Cache::Canonicalize_Expiration_Time($timespec)

Converts a timespec as described for Cache::Entry::set_expiry() into a unix
time.

=back

=cut

sub Canonicalize_Expiration_Time {
    my $timespec;

    my $timespec_param = shift(@_);
    if (! $timespec_param)
    {
        return undef;
    }
    $timespec = lc($timespec_param);

    my $time;

    if ($timespec =~ /^\s*\d+\s*$/) {
        $time = $timespec;
    }
    elsif ($timespec eq $EXPIRES_NOW) {
        $time = 0;
    }
    elsif ($timespec eq $EXPIRES_NEVER) {
        $time = undef;
    }
    elsif ($timespec =~ /^\s*-/) {
        # negative time?
        $time = 0;
    }
    elsif ($timespec =~ /^\s*\+(\d+)\s*$/) {
        $time = $1 + time();
    }
    elsif ($timespec =~ /^\s*(\+?\d+)\s*(\w*)\s*$/
        and exists($_Expiration_Units{$2}))
    {
        $time = $_Expiration_Units{$2} * $1 + time();
    }
    else {
        $time = str2time($timespec)
            or croak "invalid expiration time '$timespec'";
    }

    return $time;
}


# Hash tie methods

sub TIEHASH {
    my Cache $class = shift;
    return $class->new(@_);
}

sub STORE {
    my Cache $self = shift;
    my ($key, $value) = @_;
    return $self->set($key, $value);
}

sub FETCH {
    my Cache $self = shift;
    my ($key) = @_;
    return $self->get($key);
}

# NOT SUPPORTED
sub FIRSTKEY {
    my Cache $self = shift;
    return undef;
}

# NOT SUPPORTED
sub NEXTKEY {
    my Cache $self = shift;
    #my ($lastkey) = @_;
    return undef;
}

sub EXISTS {
    my Cache $self = shift;
    my ($key) = @_;
    return $self->exists($key);
}

sub DELETE {
    my Cache $self = shift;
    my ($key) = @_;
    return $self->remove($key);
}

sub CLEAR {
    my Cache $self = shift;
    return $self->clear();
}


1;
__END__

=head1 SEE ALSO

Cache::Entry, Cache::File, Cache::RemovalStrategy

=head1 DIFFERENCES FROM CACHE::CACHE

The Cache modules are a total redesign and reimplementation of Cache::Cache
and thus not directly compatible.  It would be, however, quite possible to
write a wrapper module that provides an identical interface to Cache::Cache.

The semantics of use are very similar to Cache::Cache, with the following
exceptions:

=over

=item The get/set methods DO NOT serialize complex data types.  Use
freeze/thaw instead (but read the notes in Cache::Entry).

=item The get_object / set_object methods are not available, but have been
superseded by the more flexible entry method and Cache::Entry class.

=item There is no concept of 'namespace' in the basic cache interface,
although implementations (eg. Cache::Memory) may choose to provide them.  For
instance, File::Cache does not provide this - but different namespaces can be
created by varying cache_root.

=item In the current Cache implementations purging is done automatically -
there is no need to explicitly enable auto purge on get/set.  The purging
algorithm is no longer implemented in the base Cache class, but is left up to
the implementations and may thus be implemented in the most efficient way for
the storage medium.

=item Cache::SharedMemory is not yet available.

=item Cache::File no longer supports separate masks for entries and
directories.  It is not a very secure configuration and presents numerous
issues for cache consistency and is hence deprecated.  There is still some
work to be done to ensure cache consistency between accesses by different
users.

=back

=head1 AUTHOR

 Chris Leishman <chris@leishman.org>
 Based on work by DeWitt Clinton <dewitt@unto.net>

=head1 COPYRIGHT

 Copyright (C) 2003-2006 Chris Leishman.  All Rights Reserved.

This module is distributed on an "AS IS" basis, WITHOUT WARRANTY OF ANY KIND,
either expressed or implied. This program is free software; you can
redistribute or modify it under the same terms as Perl itself.

$Id: Cache.pm,v 1.7 2006/01/31 15:23:58 caleishm Exp $

=cut
